client/state/data-layer/third-party/refer/index.js

```diff
@@ -38,12 +55,55 @@ const trackAffiliatePageLoad = action => {
 	);
 };
 
+const onTrackAffiliatePageLoadSuccess = ( action, response ) => {
+	if (
+		'object' !== typeof response ||
+		'object' !== typeof response.body ||
+		'object' !== typeof response.body.data
+	) {
+		debug( 'Unexpected referral response: ', response );
+		return; // Not possible.
+	}
```

Comment:
this is an appropriate place to use a `fromApi` parser. it will let us simplify the success handler and the data layer will switch to `onError` if the parse throws any exception.

```js
const onSuccess = ( action, eventProps ) => recordEvent( 'calypso_refer_visit_response', eventProps );

const fromApi = ( { body: { data: { message, status, success, ...responseData } ) => ( {
	...pick( responseData, whitelistedEventProps ),
	status: status || '',
	success: success || '',
	descriptions: message || 'success',
} );

const onError = ( action, error ) => {
	if ( ! ( error && error.response && 'string' === typeof error.response.text ) ) {
		return;
	}

	const data = JSON.parse( error.response.text );
	if ( 'object' !== typeof data ) {
		return;
	}

	return recordEvent( 'calypso_refer_visit_response', {
		...pick( data, whitelistedEventProps ),
		status: data.status || '',
		success: data.success || '',
		description: data.message || 'success',
	} );
};
```

Reply:
Really appreciate you explaining `fromApi`. What you suggest is much more elegant and I see the try/catch block switches back to `onError` as you noted. Very nice.